# Types of Classifications 

In [9]:
%pip install ucimlrepo

Note: you may need to restart the kernel to use updated packages.


In [13]:
import pandas as pd
from ucimlrepo import fetch_ucirepo

In [14]:
parkinsons_telemonitoring = fetch_ucirepo(id=189)
X = parkinsons_telemonitoring.data.features
y = parkinsons_telemonitoring.data.targets
url = 'https://archive.ics.uci.edu/static/public/189/data.csv'
df = pd.read_csv(url)
df.head()

subject#  age  sex  test_time  Jitter(%)  Jitter(Abs)  Jitter:RAP  \
0         1   72    0     5.6431    0.00662     0.000034     0.00401   
1         1   72    0    12.6660    0.00300     0.000017     0.00132   
2         1   72    0    19.6810    0.00481     0.000025     0.00205   
3         1   72    0    25.6470    0.00528     0.000027     0.00191   
4         1   72    0    33.6420    0.00335     0.000020     0.00093   

   Jitter:PPQ5  Jitter:DDP  Shimmer  ...  Shimmer:APQ5  Shimmer:APQ11  \
0      0.00317     0.01204  0.02565  ...       0.01309        0.01662   
1      0.00150     0.00395  0.02024  ...       0.01072        0.01689   
2      0.00208     0.00616  0.01675  ...       0.00844        0.01458   
3      0.00264     0.00573  0.02309  ...       0.01265        0.01963   
4      0.00130     0.00278  0.01703  ...       0.00929        0.01819   

   Shimmer:DDA       NHR     HNR     RPDE      DFA      PPE  motor_UPDRS  \
0      0.04314  0.014290  21.640  0.41888  0.54842  0.16006       28.199   
1      0.02982  0.011112  27.183  0.43493  0.56477  0.10810       28.447   
2      0.02202  0.020220  23.047  0.46222  0.54405  0.21014       28.695   
3      0.03317  0.027837  24.445  0.48730  0.57794  0.33277       28.905   
4      0.02036  0.011625  26.126  0.47188  0.56122  0.19361       29.187   

   total_UPDRS  
0       34.398  
1       34.894  
2       35.389  
3       35.810  
4       36.375  

[5 rows x 22 columns]

In [15]:
# Import necessary libraries
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics.pairwise import euclidean_distances, cosine_similarity

In [29]:
# Assuming you have loaded the dataset and named it 'df'
# Display general information about the dataset
print("Dataset Information:")
print("---------------------")
print(f"Number of Instances: {df.shape[0]}")
print(f"Number of Features: {df.shape[1]}")
print("\nData Types:")
print(df.dtypes)
print("\nMissing Values:")
print(df.isnull().sum())
print("\nUnique Values:")
print(df.nunique())

Dataset Information:
---------------------
Number of Instances: 5875
Number of Features: 22

Data Types:
subject#           int64
age                int64
sex                int64
test_time        float64
Jitter(%)        float64
Jitter(Abs)      float64
Jitter:RAP       float64
Jitter:PPQ5      float64
Jitter:DDP       float64
Shimmer          float64
Shimmer(dB)      float64
Shimmer:APQ3     float64
Shimmer:APQ5     float64
Shimmer:APQ11    float64
Shimmer:DDA      float64
NHR              float64
HNR              float64
RPDE             float64
DFA              float64
PPE              float64
motor_UPDRS      float64
total_UPDRS      float64
dtype: object

Missing Values:
subject#         0
age              0
sex              0
test_time        0
Jitter(%)        0
Jitter(Abs)      0
Jitter:RAP       0
Jitter:PPQ5      0
Jitter:DDP       0
Shimmer          0
Shimmer(dB)      0
Shimmer:APQ3     0
Shimmer:APQ5     0
Shimmer:APQ11    0
Shimmer:DDA      0
NHR              0
HNR       

In [10]:
print("\nSummary Statistics:")
print(df.describe())


Summary Statistics:
          subject#          age          sex    test_time    Jitter(%)  \
count  5875.000000  5875.000000  5875.000000  5875.000000  5875.000000   
mean     21.494128    64.804936     0.317787    92.863722     0.006154   
std      12.372279     8.821524     0.465656    53.445602     0.005624   
min       1.000000    36.000000     0.000000    -4.262500     0.000830   
25%      10.000000    58.000000     0.000000    46.847500     0.003580   
50%      22.000000    65.000000     0.000000    91.523000     0.004900   
75%      33.000000    72.000000     1.000000   138.445000     0.006800   
max      42.000000    85.000000     1.000000   215.490000     0.099990   

       Jitter(Abs)   Jitter:RAP  Jitter:PPQ5   Jitter:DDP      Shimmer  ...  \
count  5875.000000  5875.000000  5875.000000  5875.000000  5875.000000  ...   
mean      0.000044     0.002987     0.003277     0.008962     0.034035  ...   
std       0.000036     0.003124     0.003732     0.009371     0.025835  ...

In [7]:
# Display information about the target variable ('motor_UPDRS')
target_variable = 'motor_UPDRS'


In [9]:
# Formulate binary classification problem
threshold = 20  # Choose an appropriate threshold
df['target_class'] = (df[target_variable] >= threshold).astype(int)


In [10]:
# Display information about the target variable (assuming 'y' is your target variable)
print("\nTarget Variable Information:")
print("-----------------------------")
print(f"Number of Unique Classes: {df['target_class'].nunique()}")
print("\nClass Distribution:")
print(df['target_class'].value_counts())


Target Variable Information:
-----------------------------
Number of Unique Classes: 2

Class Distribution:
target_class
1    3115
0    2760
Name: count, dtype: int64


### Probabilistic Classifiier Code

In [25]:
# Probabilistic Classifier
prob_classifier = GaussianNB()
prob_classifier.fit(X_train, y_train)
y_prob_pred = prob_classifier.predict(X_test)
confusion_mat_prob = confusion_matrix(y_test, y_prob_pred)

###  Probabilistic Classification Results

In [16]:
# Probabilistic Classification Results
confusion_mat_prob = confusion_matrix(y_test, y_prob_pred)
print("\nProbabilistic Classification Results:")
print("Confusion Matrix:\n", confusion_mat_prob)


Probabilistic Classification Results:
Confusion Matrix:
 [[524  39]
 [246 366]]


Probabilistic Classification System:



Explanation:


The probabilistic classifier predicted 524 instances correctly as 'Class 0' and 366 instances correctly as 'Class 1'.

However, there were 246 misclassifications of 'Class 0' as 'Class 1' and 39 misclassifications of 'Class 1' as 'Class 0'.

The probabilistic classifier, based on Gaussian Naive Bayes, demonstrated reasonable performance.

True Positive (TP): 366 patients correctly classified as having higher motor_UPDRS (>=20).

True Negative (TN): 524 patients correctly classified as having lower motor_UPDRS (<20).

False Positive (FP): 39 patients misclassified as having higher motor_UPDRS.

False Negative (FN): 246 patients misclassified as having lower motor_UPDRS.




### Euclidean Distance Classifier & Euclidean Distance Classification Results

Euclidean Distance Classification System:

Implementation:

Calculated Euclidean distances between test set samples and training set samples.
Assigned the class label of the nearest neighbor in the training set.

Results:

Produced a confusion matrix (confusion_mat_euclidean) to assess the accuracy and misclassifications of the Euclidean distance classifier.

In [17]:
# Euclidean Distance Classifier
euclidean_X_train = X_train
euclidean_y_train = y_train
distances_euclidean = euclidean_distances(X_test, euclidean_X_train)

In [18]:
# Ensure indices are within the range of euclidean_y_train
valid_indices = [i for i in distances_euclidean.argmin(axis=1) if i < len(euclidean_y_train)]
y_euclidean_pred = [euclidean_y_train.iloc[i] for i in valid_indices]


In [19]:
confusion_mat_euclidean = confusion_matrix(y_test, y_euclidean_pred)
print("\nEuclidean Distance Classification Results:")
print("Confusion Matrix:\n", confusion_mat_euclidean)


Euclidean Distance Classification Results:
Confusion Matrix:
 [[563   0]
 [  0 612]]


Explanation:

The Euclidean Distance classifier achieved perfect accuracy.
All test samples were correctly classified with no misclassifications.
This could indicate a clear separation in the feature space, making samples from different classes easily distinguishable.


### Cosine Similarity Classifier<br>

   ### Cosine Similarity Classification Results


In [26]:
# Cosine Similarity Classifier
cosine_X_train = X_train
cosine_y_train = y_train
similarities_cosine = cosine_similarity(X_test, cosine_X_train)


In [27]:
# Ensure indices are within the range of cosine_y_train
valid_indices_cosine = [i for i in similarities_cosine.argmax(axis=1) if i < len(cosine_y_train)]
y_cosine_pred = [cosine_y_train.iloc[i] for i in valid_indices_cosine]

confusion_mat_cosine = confusion_matrix(y_test, y_cosine_pred)
print("\nCosine Similarity Classification Results:")
print("Confusion Matrix:\n", confusion_mat_cosine)


Cosine Similarity Classification Results:
Confusion Matrix:
 [[559   4]
 [  2 610]]


The Cosine Similarity classifier had a few misclassifications, with 4 instances of 'Class 0' incorrectly predicted as 'Class 1' and 2 instances of 'Class 1' incorrectly predicted as 'Class 0'.


In [16]:
conf_matrix = np.array([[559, 4], [2, 610]])

def calculate_accuracy(confusion_matrix):
    true_positives = confusion_matrix[0, 0]
    true_negatives = confusion_matrix[1, 1]
    total_population = confusion_matrix.sum()

    accuracy = (true_positives + true_negatives) / total_population
    return accuracy

accuracy = calculate_accuracy(conf_matrix)
print(f'Accuracy: {accuracy * 100:.2f}%')


Accuracy: 99.49%
